### PR TITLE
Weapon lights

### PR DESCRIPTION
**Pull request by EatThePath**
Configurable lighting data for weapons, both on a per-weapon basis and globally via lighting profiles. 

I will be doing a bit more testing to makes sure last minute changes or history rewriting didn't break anything, but this is now ready to review.

### CODE REVIEW

**Comment by z64555** [comment]
For every 'new' there must be a 'delete' at the end of the lifetime of the created object.  'delete' is usually written in the destructor of the class that called 'new'

Member data that are standard data types char, int, float, etc. and arrays do not need to be explicity cleaned up in the destructor of a class, they will be dealt with automatically.

Member data that are pointers to objects do not need to be managed either, however their referenced data must be cleaned up one way or another.  Objects that are defined in the codebase "normally" (that is, *not* using 'new' or 'malloc') will be cleaned up the moment they go out of execution scope. (Lookup scope for C++)

<hr>
Make use of the keywords 'public', 'protected', and 'private' to control access to methods and member data. Public is accessible to all. Protected is accessible to the class and its children, private is accessible only to itself.  Class friendship is also a thing but is usually not useful.

When prototyping a new class its easy to  just slap down 'public' and get to coding, but this is a bad habit.  You should think at the start of how its members should be interacted with the rest of the codebase.
  Public member data can be both accessed and modified by all sources, which can cause action at a distance bugs due to abuse of public members, but it does ease the workload of needing to supply getter and setter methods to access member data.

Protected members and methods can be accessed and modified by the class and its children/subclasses that inherit it.  This should be the highest level of helper subroutines.

Refrain from defining helper subroutines C-style within the .cpp implementation of the class, as this obfuscates them from maintainers and means they're effectively private static methods.

Private members and methods are data that can *only* be used by that class. If member data are protected or private they must have getter and setter methods for outside code to see or modify them.  Pointers to data on the heap (from 'new' or 'malloc') that are created within the class should be 'protected' or 'private' to gate potential abuse.

**Comment by wookieejedi** [comment]
Great to see this one is ready for review, really looking forward to being able to use these updates in FotG (and I'm sure other mods are too). Given the smaller number of graphics coders, I'm guessing @TRBlount would be most qualified to start this off?

**Comment by EatThePath** [comment]
> Great to see this one is ready for review, really looking forward to being able to use these updates in FotG (and I'm sure other mods are too). Given the smaller number of graphics coders, I'm guessing @TRBlount would be most qualified to start this off?

Possibly yes. Fiddling with weapons and beams brings @Baezon to mind, and the parsing sections makes me think @JohnAFernandez too. Ultimately I don't thinking too much of this is sticky code, but it touches enough things to be a bit of a task to gather up all the context I guess.

File: code/parse/parsehi.cpp
Comment (Goober5000) [inline code/parse/parsehi.cpp:55]: This isn't going to do anything for `out_color` because the calling function won't have the address of the allocated memory.  To make this work you would need a pointer-to-pointer.  However, given the risky memory management here, I would recommend taking this out and just adding an Assertion that the pointer isn't null.

File: code/weapon/weapon.h
Comment (Goober5000) [inline code/weapon/weapon.h:256]: Why not add these to the flag set?

File: code/weapon/weapon.h
Comment (Goober5000) [inline code/weapon/weapon.h:360]: Similarly, this could be added to the weapon_info flag set

File: code/object/object.cpp
Comment (Goober5000) [inline code/object/object.cpp:1217]: Why did you move this outside the if() block?

File: code/graphics/color.cpp
Comment (Goober5000) [inline code/graphics/color.cpp:7]: You could make this a const reference (`const SCP_vector<float>&`) if you want

File: code/graphics/color.cpp
Comment (Goober5000) [inline code/graphics/color.cpp:26]: You could make this a non-const reference.  In fact, this has the same pointer bug and memory risk as the function in parsehi.

File: code/lighting/lighting_profiles.cpp
Comment (Goober5000) [inline code/lighting/lighting_profiles.cpp:121]: "too man no"?

File: code/lighting/lighting_profiles.cpp
Comment (Goober5000) [inline code/lighting/lighting_profiles.cpp:262]: I would suggest setting a `const char *` variable to "Default profile" and then using that for each line.

**Review by Goober5000 — CHANGES_REQUESTED** [review]
Reviewed a lot of it but skipped the graphics/lighting code.  Noticed a few things.

File: code/weapon/weapon.h
Comment (EatThePath) [reply code/weapon/weapon.h:256]: mainly I forgot flagsets exist. Will correct.

File: code/object/object.cpp
Comment (EatThePath) [reply code/object/object.cpp:1217]: I believe I was at one intermediate stage using something from wi in the if condition, and forgot to moved it back when that became no-longer true. I'll put it back

File: code/lighting/lighting_profiles.cpp
Comment (EatThePath) [reply code/lighting/lighting_profiles.cpp:121]: I wonder how tired I was when writing that one, woof.

File: code/weapon/weapon.h
Comment (EatThePath) [reply code/weapon/weapon.h:256]: Upon review and discussion, adding these(and the other bool later) to the existing regular flag sets would have undesirable side effects. To my mind creating new flagsets for them would be added complexity for little to no benefit, so I'm inclined to leave them as is unless there's a compelling benefit I'm unaware of.
The other review items have been addressed, though.

**Comment by Goober5000** [comment]
Ok, LGTM now for the parts I looked at.  Still need someone else to review the lighting code.

File: code/weapon/weapon.h
Comment (Baezon) [reply code/weapon/weapon.h:256]: What undesirable side effects would there be to adding it to beam info flags?

Comment (EatThePath) [reply code/weapon/weapon.h:256]: It was my understanding that if anyone adds a beam flag in a tbm it will clear any beam flags set upstream, which isn't something I much care for really. Even if I was misinformed or misunderstood and that doesn't apply here, I'd still be curious to know what benefit that system provides over this method.

File: code/graphics/color.h
Comment (TRBlount) [inline code/graphics/color.h:1]: Generally, we seem to use `#ifndef _FILE_H_` instead of `#pragma once`, would need feedback from others on how they feel about this.

File: code/lighting/lighting.cpp
Comment (TRBlount) [inline code/lighting/lighting.cpp:156]: Several function calls don't match a visible prototype/definition here.

File: code/weapon/weapon.h
Comment (TRBlount) [reply code/weapon/weapon.h:256]: Is this resolved or awaiting followup?

**Review by TRBlount — COMMENTED** [review]
Waiting for reviewer feedback and some fixes.

File: code/weapon/weapon.h
Comment (EatThePath) [reply code/weapon/weapon.h:256]: @TRBlount I took it as resolved, personally, but I could be wrong

Comment (Goober5000) [reply code/weapon/weapon.h:256]: It's resolved.  The only existing flagset is the weapon flags, which is not appropriate for this.

File: code/lighting/lighting.cpp
Comment (EatThePath) [reply code/lighting/lighting.cpp]: Copy-paste oversights. I think I got them all in my last push

**Review by TRBlount — APPROVED** [review]
Looks good now.

**Review by Goober5000 — APPROVED** [review]

**Pull request merged**
